[java] AssertionError: Method should be accessible

**Affects PMD Version:** 

7.0.0

**Description:**

Errors about unreachable methods are intermittently (only 50% of the times) produced.

1. Unknown error occurred while executing a PmdRunnable: net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTStringLiteral.getLiteralText()Chars/invokeVirtual
Exception Context:
        [1:Rule applied on node=!debug only! [StringLiteral:9:29]"*retracted*"]
          at net.sourceforge.pmd.util.AssertionUtil.shouldNotReachHere(AssertionUtil.java:219)
2. Unknown error occurred while executing a PmdRunnable: net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTNumericLiteral.getLiteralText()Chars/invokeVirtual
Exception Context:
        [1:Rule applied on node=!debug only! [NumericLiteral:11:21]0]
3. Unknown error occurred while executing a PmdRunnable: net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTStringLiteral.getLiteralText()Chars/invokeVirtual
Exception Context:
        [1:Rule applied on node=!debug only! [StringLiteral:15:31]"no classes should use field injection"]

**Exception Stacktrace:**

1. ```org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':<retracted>:lib:archunit-rules:pmdMain'.
            at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:130)
           ....
    Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing org.gradle.api.plugins.quality.internal.PmdAction
            at org.gradle.workers.internal.DefaultWorkerExecutor$WorkItemExecution.waitForCompletion(DefaultWorkerExecutor.java:287)
            ....
    Caused by: : net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTStringLiteral.getLiteralText()Chars/invokeVirtual
    Exception Context:
            [1:Rule applied on node=!debug only! [StringLiteral:9:29]"<retracted>"]
    ---------------------------------
            at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:109)
            ....
    Caused by: net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTStringLiteral.getLiteralText()Chars/invokeVirtual
    Exception Context:
            [1:Rule applied on node=!debug only! [StringLiteral:9:29]"<retracted>"]
    ---------------------------------
            at net.sourceforge.pmd.util.AssertionUtil.shouldNotReachHere(AssertionUtil.java:219)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.lambda$getWrappersForClass$1(AttributeAxisIterator.java:92)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.getWrappersForClass(AttributeAxisIterator.java:95)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.<init>(AttributeAxisIterator.java:82)
            at net.sourceforge.pmd.lang.ast.Node.getXPathAttributesIterator(Node.java:278)
            at net.sourceforge.pmd.lang.rule.xpath.internal.AstElementNode.getLightAttributes(AstElementNode.java:105)
            at net.sourceforge.pmd.lang.rule.xpath.internal.AstElementNode.getAttributeValue(AstElementNode.java:178)
            at net.sf.saxon.expr.AttributeGetter.evaluateItem(AttributeGetter.java:127)
            at net.sf.saxon.expr.AtomicSequenceConverter.evaluateItem(AtomicSequenceConverter.java:364)
            at net.sf.saxon.expr.AtomicSequenceConverter.evaluateItem(AtomicSequenceConverter.java:28)
            at net.sf.saxon.expr.Expression.evaluateAsString(Expression.java:917)
            at net.sf.saxon.functions.StartsWith$1.effectiveBooleanValue(StartsWith.java:52)
            at net.sf.saxon.expr.FilterIterator$NonNumeric.matches(FilterIterator.java:180)
            at net.sf.saxon.expr.FilterIterator.getNextMatchingItem(FilterIterator.java:77)
            at net.sf.saxon.expr.FilterIterator.next(FilterIterator.java:63)
            at net.sourceforge.pmd.lang.rule.xpath.internal.SaxonXPathRuleQuery.evaluate(SaxonXPathRuleQuery.java:127)
            at net.sourceforge.pmd.lang.rule.xpath.XPathRule.apply(XPathRule.java:101)
            at net.sourceforge.pmd.lang.rule.RuleReference.apply(RuleReference.java:415)
            at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex(RuleApplicator.java:77)
            at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply(RuleApplicator.java:57)
            at net.sourceforge.pmd.lang.rule.internal.RuleSets.apply(RuleSets.java:155)
            at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource(PmdRunnable.java:140)
            at net.sourceforge.pmd.lang.impl.PmdRunnable.run(PmdRunnable.java:80)
    ```
2. ```1   org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':<retracted>:lib:archunit-rules:pmdRuleViolationsTest'.
            at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:130)
            ....
    Caused by: : net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTNumericLiteral.getLiteralText()Chars/invokeVirtual
    Exception Context:
            [1:Rule applied on node=!debug only! [NumericLiteral:11:21]0]
    ---------------------------------
            at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:109)
           ....
    Caused by: net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTNumericLiteral.getLiteralText()Chars/invokeVirtual
    Exception Context:
            [1:Rule applied on node=!debug only! [NumericLiteral:11:21]0]
    ---------------------------------
            at net.sourceforge.pmd.util.AssertionUtil.shouldNotReachHere(AssertionUtil.java:219)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.lambda$getWrappersForClass$1(AttributeAxisIterator.java:92)
            at java.base@11.0.22/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
            at java.base@11.0.22/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
            at java.base@11.0.22/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
            at java.base@11.0.22/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
            at java.base@11.0.22/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
            at java.base@11.0.22/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
            at java.base@11.0.22/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
            at java.base@11.0.22/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.getWrappersForClass(AttributeAxisIterator.java:95)
            at java.base@11.0.22/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.<init>(AttributeAxisIterator.java:82)
            at net.sourceforge.pmd.lang.ast.Node.getXPathAttributesIterator(Node.java:278)
            at net.sourceforge.pmd.lang.rule.xpath.internal.AstElementNode.getLightAttributes(AstElementNode.java:105)
            at net.sourceforge.pmd.lang.rule.xpath.internal.AstElementNode.getAttributeValue(AstElementNode.java:178)
            at net.sf.saxon.expr.AttributeGetter.evaluateItem(AttributeGetter.java:127)
            at net.sf.saxon.expr.ValueComparison.effectiveBooleanValue(ValueComparison.java:357)
            at net.sf.saxon.expr.FilterIterator$NonNumeric.matches(FilterIterator.java:180)
            at net.sf.saxon.expr.FilterIterator.getNextMatchingItem(FilterIterator.java:77)
            at net.sf.saxon.expr.FilterIterator.next(FilterIterator.java:63)
            at net.sf.saxon.om.FocusTrackingIterator.next(FocusTrackingIterator.java:76)
            at net.sf.saxon.expr.FilterIterator.getNextMatchingItem(FilterIterator.java:76)
            at net.sf.saxon.expr.FilterIterator.next(FilterIterator.java:63)
            at net.sf.saxon.om.FocusTrackingIterator.next(FocusTrackingIterator.java:76)
            at net.sf.saxon.expr.FilterIterator.getNextMatchingItem(FilterIterator.java:76)
            at net.sf.saxon.expr.FilterIterator.next(FilterIterator.java:63)
            at net.sourceforge.pmd.lang.rule.xpath.internal.SaxonXPathRuleQuery.evaluate(SaxonXPathRuleQuery.java:127)
            at net.sourceforge.pmd.lang.rule.xpath.XPathRule.apply(XPathRule.java:101)
            at net.sourceforge.pmd.lang.rule.RuleReference.apply(RuleReference.java:415)
            at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex(RuleApplicator.java:77)
            at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply(RuleApplicator.java:57)
            at net.sourceforge.pmd.lang.rule.internal.RuleSets.apply(RuleSets.java:155)
            at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource(PmdRunnable.java:140)
            at net.sourceforge.pmd.lang.impl.PmdRunnable.run(PmdRunnable.java:80)
            at java.base@11.0.22/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
            at java.base@11.0.22/java.util.concurrent.FutureTask.run(FutureTask.java:264)
            ... 3 more```
3. see next comment, can't post here due to character limit


**Code Sample demonstrating the issue:**

I'm not a 100% sure, but these lines of code seem to be involved in producing the errors.

1.
```java
    @AnalyzeClasses(packages = {"<retracted-package-1>", "<retracted-package-2>"})
```
2. 
```java
    assertNotEquals(0, hashCode());
```
3. 
```java
    assertThatThrownBy(() -> MainArchUnitTest.NO_CLASSES_SHOULD_USE_FIELD_INJECTION.check(jc)).hasMessageContaining("no classes should use field injection");
```

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1.  `./gradlew clean build --stacktrace`

**Running PMD through:** *Gradle*

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->
JDK 11

Third exception:
```org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':<retracted>:lib:archunit-rules:pmdTest'.
            at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:130)
            at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:282)
            at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:128)
            at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:116)
            at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
            at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
            at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
            at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:74)
            at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
            at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
            at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
            at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
            at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
            at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
            at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:42)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:331)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:318)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.lambda$execute$0(DefaultTaskExecutionGraph.java:314)
            at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:80)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:314)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:303)
            at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:463)
            at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:380)
            at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
            at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
            at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
            at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
            at java.base/java.lang.Thread.run(Thread.java:829)
    Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing org.gradle.api.plugins.quality.internal.PmdAction
            at org.gradle.workers.internal.DefaultWorkerExecutor$WorkItemExecution.waitForCompletion(DefaultWorkerExecutor.java:287)
            at org.gradle.internal.work.DefaultAsyncWorkTracker.lambda$waitForItemsAndGatherFailures$2(DefaultAsyncWorkTracker.java:130)
            at org.gradle.internal.Factories$1.create(Factories.java:31)
            at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLocks(DefaultWorkerLeaseService.java:336)
            at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLocks(DefaultWorkerLeaseService.java:319)
            at org.gradle.internal.work.DefaultWorkerLeaseService.withoutLock(DefaultWorkerLeaseService.java:324)
            at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:126)
            at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:92)
            at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForAll(DefaultAsyncWorkTracker.java:78)
            at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForCompletion(DefaultAsyncWorkTracker.java:66)
            at org.gradle.api.internal.tasks.execution.TaskExecution$3.run(TaskExecution.java:252)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
            at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
            at org.gradle.api.internal.tasks.execution.TaskExecution.executeAction(TaskExecution.java:229)
            at org.gradle.api.internal.tasks.execution.TaskExecution.executeActions(TaskExecution.java:212)
            at org.gradle.api.internal.tasks.execution.TaskExecution.executeWithPreviousOutputFiles(TaskExecution.java:195)
            at org.gradle.api.internal.tasks.execution.TaskExecution.execute(TaskExecution.java:162)
            at org.gradle.internal.execution.steps.ExecuteStep.executeInternal(ExecuteStep.java:105)
            at org.gradle.internal.execution.steps.ExecuteStep.access$000(ExecuteStep.java:44)
            at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:59)
            at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:56)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
            at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
            at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:56)
            at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:44)
            at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:41)
            at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:74)
            at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:55)
            at org.gradle.internal.execution.steps.PreCreateOutputParentsStep.execute(PreCreateOutputParentsStep.java:50)
            at org.gradle.internal.execution.steps.PreCreateOutputParentsStep.execute(PreCreateOutputParentsStep.java:28)
            at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:67)
            at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:37)
            at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:61)
            at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:26)
            at org.gradle.internal.execution.steps.CaptureOutputsAfterExecutionStep.execute(CaptureOutputsAfterExecutionStep.java:67)
            at org.gradle.internal.execution.steps.CaptureOutputsAfterExecutionStep.execute(CaptureOutputsAfterExecutionStep.java:45)
            at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:40)
            at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:29)
            at org.gradle.internal.execution.steps.BuildCacheStep.executeWithoutCache(BuildCacheStep.java:189)
            at org.gradle.internal.execution.steps.BuildCacheStep.lambda$execute$1(BuildCacheStep.java:75)
            at org.gradle.internal.Either$Right.fold(Either.java:175)
            at org.gradle.internal.execution.caching.CachingState.fold(CachingState.java:62)
            at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:73)
            at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:48)
            at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:46)
            at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:35)
            at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:76)
            at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$2(SkipUpToDateStep.java:54)
            at java.base/java.util.Optional.orElseGet(Optional.java:369)
            at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:54)
            at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:36)
            at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:37)
            at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:27)
            at org.gradle.internal.execution.steps.ResolveIncrementalCachingStateStep.executeDelegate(ResolveIncrementalCachingStateStep.java:49)
            at org.gradle.internal.execution.steps.ResolveIncrementalCachingStateStep.executeDelegate(ResolveIncrementalCachingStateStep.java:27)
            at org.gradle.internal.execution.steps.AbstractResolveCachingStateStep.execute(AbstractResolveCachingStateStep.java:71)
            at org.gradle.internal.execution.steps.AbstractResolveCachingStateStep.execute(AbstractResolveCachingStateStep.java:39)
            at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:65)
            at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:36)
            at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:106)
            at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:55)
            at org.gradle.internal.execution.steps.AbstractCaptureStateBeforeExecutionStep.execute(AbstractCaptureStateBeforeExecutionStep.java:64)
            at org.gradle.internal.execution.steps.AbstractCaptureStateBeforeExecutionStep.execute(AbstractCaptureStateBeforeExecutionStep.java:43)
            at org.gradle.internal.execution.steps.AbstractSkipEmptyWorkStep.executeWithNonEmptySources(AbstractSkipEmptyWorkStep.java:125)
            at org.gradle.internal.execution.steps.AbstractSkipEmptyWorkStep.execute(AbstractSkipEmptyWorkStep.java:61)
            at org.gradle.internal.execution.steps.AbstractSkipEmptyWorkStep.execute(AbstractSkipEmptyWorkStep.java:36)
            at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
            at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:36)
            at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:23)
            at org.gradle.internal.execution.steps.HandleStaleOutputsStep.execute(HandleStaleOutputsStep.java:75)
            at org.gradle.internal.execution.steps.HandleStaleOutputsStep.execute(HandleStaleOutputsStep.java:41)
            at org.gradle.internal.execution.steps.AssignMutableWorkspaceStep.lambda$execute$0(AssignMutableWorkspaceStep.java:35)
            at org.gradle.api.internal.tasks.execution.TaskExecution$4.withWorkspace(TaskExecution.java:289)
            at org.gradle.internal.execution.steps.AssignMutableWorkspaceStep.execute(AssignMutableWorkspaceStep.java:31)
            at org.gradle.internal.execution.steps.AssignMutableWorkspaceStep.execute(AssignMutableWorkspaceStep.java:22)
            at org.gradle.internal.execution.steps.ChoosePipelineStep.execute(ChoosePipelineStep.java:40)
            at org.gradle.internal.execution.steps.ChoosePipelineStep.execute(ChoosePipelineStep.java:23)
            at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.lambda$execute$2(ExecuteWorkBuildOperationFiringStep.java:67)
            at java.base/java.util.Optional.orElseGet(Optional.java:369)
            at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.execute(ExecuteWorkBuildOperationFiringStep.java:67)
            at org.gradle.internal.execution.steps.ExecuteWorkBuildOperationFiringStep.execute(ExecuteWorkBuildOperationFiringStep.java:39)
            at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:46)
            at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:34)
            at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:48)
            at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:35)
            at org.gradle.internal.execution.impl.DefaultExecutionEngine$1.execute(DefaultExecutionEngine.java:61)
            at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:127)
            at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:116)
            at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
            at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
            at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
            at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:74)
            at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
            at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
            at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
            at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
            at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
            at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
            at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
            at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
            at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:42)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:331)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:318)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.lambda$execute$0(DefaultTaskExecutionGraph.java:314)
            at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:80)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:314)
            at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:303)
            at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:463)
            at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:380)
            at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
            at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
            at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
            at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
            at java.base/java.lang.Thread.run(Thread.java:829)
    Caused by: : net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTStringLiteral.getLiteralText()Chars/invokeVirtual
    Exception Context:
            [1:Rule applied on node=!debug only! [StringLiteral:15:31]"no classes should use field injection"]
    ---------------------------------
            at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:109)
            at groovy.ant.AntBuilder.performTask(AntBuilder.java:343)
            at groovy.ant.AntBuilder.nodeCompleted(AntBuilder.java:282)
            at org.gradle.api.internal.project.ant.BasicAntBuilder.nodeCompleted(BasicAntBuilder.java:90)
            at jdk.internal.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
            at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
            at java.base/java.lang.reflect.Method.invoke(Method.java:566)
            at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
            at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
            at org.gradle.internal.metaobject.BeanDynamicObject$MetaClassAdapter.invokeMethod(BeanDynamicObject.java:541)
            at org.gradle.internal.metaobject.BeanDynamicObject.tryInvokeMethod(BeanDynamicObject.java:218)
            at org.gradle.internal.metaobject.AbstractDynamicObject.invokeMethod(AbstractDynamicObject.java:163)
            at org.gradle.api.internal.project.antbuilder.AntBuilderDelegate.nodeCompleted(AntBuilderDelegate.java:124)
            at groovy.util.BuilderSupport.doInvokeMethod(BuilderSupport.java:160)
            at groovy.util.BuilderSupport.invokeMethod(BuilderSupport.java:74)
            at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:44)
            at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)
            at org.gradle.api.plugins.quality.internal.PmdInvoker.runPmd(PmdInvoker.groovy:120)
            at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
            at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
            at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
            at java.base/java.lang.reflect.Method.invoke(Method.java:566)
            at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
            at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
            at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1258)
            at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodOnCurrentN(ScriptBytecodeAdapter.java:94)
            at org.gradle.api.plugins.quality.internal.PmdInvoker.this$dist$invoke$1(PmdInvoker.groovy)
            at org.gradle.api.plugins.quality.internal.PmdInvoker$1.methodMissing(PmdInvoker.groovy)
            at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
            at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
            at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
            at java.base/java.lang.reflect.Method.invoke(Method.java:566)
            at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
            at groovy.lang.MetaClassImpl.invokeMissingMethod(MetaClassImpl.java:953)
            at groovy.lang.MetaClassImpl.invokePropertyOrMissing(MetaClassImpl.java:1337)
            at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1260)
            at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:61)
            at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:194)
            at org.gradle.api.plugins.quality.internal.PmdInvoker$1.create(PmdInvoker.groovy:56)
            at org.gradle.api.plugins.quality.internal.PmdInvoker$1.create(PmdInvoker.groovy)
            at org.gradle.internal.SystemProperties.withSystemProperty(SystemProperties.java:122)
            at org.gradle.internal.SystemProperties$withSystemProperty.call(Unknown Source)
            at org.gradle.api.plugins.quality.internal.PmdInvoker.execute(PmdInvoker.groovy:53)
            at org.gradle.api.plugins.quality.internal.PmdInvoker.execute(PmdInvoker.groovy)
            at org.gradle.api.internal.project.antbuilder.DefaultIsolatedAntBuilder.lambda$execute$2(DefaultIsolatedAntBuilder.java:158)
            at org.gradle.api.internal.project.antbuilder.ClassPathToClassLoaderCache.withCachedClassLoader(ClassPathToClassLoaderCache.java:135)
            at org.gradle.api.internal.project.antbuilder.DefaultIsolatedAntBuilder.execute(DefaultIsolatedAntBuilder.java:141)
            at org.gradle.api.plugins.quality.internal.AntWorkAction.execute(AntWorkAction.java:38)
            at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
            at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:54)
            at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:48)
            at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:100)
            at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:48)
            at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
            at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
            at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:102)
            at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:71)
            at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:146)
            at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
            at org.gradle.process.internal.worker.request.WorkerAction.lambda$run$0(WorkerAction.java:143)
            at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:80)
            at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
            at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
            at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
            at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
            at java.base/java.lang.reflect.Method.invoke(Method.java:566)
            at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
            at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
            at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
            at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
            at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
            at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
            at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
            ... 3 more
    Caused by: net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method should be accessible java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTStringLiteral.getLiteralText()Chars/invokeVirtual
    Exception Context:
            [1:Rule applied on node=!debug only! [StringLiteral:15:31]"no classes should use field injection"]
    ---------------------------------
            at net.sourceforge.pmd.util.AssertionUtil.shouldNotReachHere(AssertionUtil.java:219)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.lambda$getWrappersForClass$1(AttributeAxisIterator.java:92)
            at java.base@11.0.22/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
            at java.base@11.0.22/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
            at java.base@11.0.22/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
            at java.base@11.0.22/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
            at java.base@11.0.22/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
            at java.base@11.0.22/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
            at java.base@11.0.22/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
            at java.base@11.0.22/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.getWrappersForClass(AttributeAxisIterator.java:95)
            at java.base@11.0.22/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1737)
            at net.sourceforge.pmd.lang.rule.xpath.impl.AttributeAxisIterator.<init>(AttributeAxisIterator.java:82)
            at net.sourceforge.pmd.lang.ast.Node.getXPathAttributesIterator(Node.java:278)
            at net.sourceforge.pmd.lang.rule.xpath.internal.AstElementNode.getLightAttributes(AstElementNode.java:105)
            at net.sourceforge.pmd.lang.rule.xpath.internal.AstElementNode.getAttributeValue(AstElementNode.java:178)
            at net.sf.saxon.expr.AttributeGetter.evaluateItem(AttributeGetter.java:127)
            at net.sf.saxon.expr.AtomicSequenceConverter.evaluateItem(AtomicSequenceConverter.java:364)
            at net.sf.saxon.expr.AtomicSequenceConverter.evaluateItem(AtomicSequenceConverter.java:28)
            at net.sf.saxon.expr.Expression.evaluateAsString(Expression.java:917)
            at net.sf.saxon.functions.StartsWith$1.effectiveBooleanValue(StartsWith.java:52)
            at net.sf.saxon.expr.FilterIterator$NonNumeric.matches(FilterIterator.java:180)
            at net.sf.saxon.expr.FilterIterator.getNextMatchingItem(FilterIterator.java:77)
            at net.sf.saxon.expr.FilterIterator.next(FilterIterator.java:63)
            at net.sourceforge.pmd.lang.rule.xpath.internal.SaxonXPathRuleQuery.evaluate(SaxonXPathRuleQuery.java:127)
            at net.sourceforge.pmd.lang.rule.xpath.XPathRule.apply(XPathRule.java:101)
            at net.sourceforge.pmd.lang.rule.RuleReference.apply(RuleReference.java:415)
            at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex(RuleApplicator.java:77)
            at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply(RuleApplicator.java:57)
            at net.sourceforge.pmd.lang.rule.internal.RuleSets.apply(RuleSets.java:155)
            at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource(PmdRunnable.java:140)
            at net.sourceforge.pmd.lang.impl.PmdRunnable.run(PmdRunnable.java:80)
            at java.base@11.0.22/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
            at java.base@11.0.22/java.util.concurrent.FutureTask.run(FutureTask.java:264)
            ... 3 more
```

```
Execution failed for task ':core:pmdMain'.
> net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method 'public final net.sourceforge.pmd.lang.document.Chars net.sourceforge.pmd.lang.java.ast.AbstractLiteral.getLiteralText()' should be accessible, but: java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTNumericLiteral.getLiteralText()Chars/invokeVirtual
  Exception Context:
        [1:Rule applied on node=!debug only! [NumericLiteral:32:44]0]
```

This is with PMD 7.2.0

Another one:

```
> net.sourceforge.pmd.util.ContextedAssertionError: This should be unreachable: Method 'public final net.sourceforge.pmd.lang.document.Chars net.sourceforge.pmd.lang.java.ast.AbstractLiteral.getLiteralText()' should be accessible, but: java.lang.IllegalAccessException: no such method: net.sourceforge.pmd.lang.java.ast.ASTStringLiteral.getLiteralText()Chars/invokeVirtual
  Exception Context:
        [1:Rule applied on node=!debug only! [StringLiteral:35:38]"«redacted»"]
```

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -17,6 +17,8 @@ This is a {{ site.pmd.release_type }} release.
 ### 🐛 Fixed Issues
 * apex
   * [#5053](https://github.com/pmd/pmd/issues/5053): \[apex] CPD fails to parse string literals with escaped characters
+* java
+  * [#4885](https://github.com/pmd/pmd/issues/4885): \[java] AssertionError: Method should be accessible
 * java-bestpractices
   * [#5047](https://github.com/pmd/pmd/issues/5047): \[java] UnusedPrivateMethod FP for Generics & Overloads
 * plsql

--- a/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/impl/AttributeAxisIterator.java
+++ b/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/impl/AttributeAxisIterator.java
@@ -4,7 +4,6 @@
 
 package net.sourceforge.pmd.lang.rule.xpath.impl;
 
-import static net.sourceforge.pmd.util.CollectionUtil.emptyList;
 import static net.sourceforge.pmd.util.CollectionUtil.setOf;
 
 import java.lang.invoke.MethodHandle;
@@ -91,7 +90,7 @@ public class AttributeAxisIterator implements Iterator<Attribute> {
                      .filter(m -> isAttributeAccessor(nodeClass, m))
                      .map(m -> {
                          try {
-                             return new MethodWrapper(m, nodeClass);
+                             return new MethodWrapper(m);
                          } catch (ReflectiveOperationException e) {
                              throw AssertionUtil.shouldNotReachHere("Method '" + m + "' should be accessible, but: " + e, e);
                          }
@@ -210,19 +209,13 @@ public class AttributeAxisIterator implements Iterator<Attribute> {
         public final String name;
 
 
-        MethodWrapper(Method m, Class<?> nodeClass) throws IllegalAccessException, NoSuchMethodException {
+        MethodWrapper(Method m) throws IllegalAccessException {
             this.method = m;
             this.name = truncateMethodName(m.getName());
-
-            if (!Modifier.isPublic(m.getDeclaringClass().getModifiers())) {
-                // This is a public method of a non-public class.
-                // To call it from reflection we need to call it via invokevirtual,
-                // whereas the default handle would use invokespecial.
-                MethodType methodType = MethodType.methodType(m.getReturnType(), emptyList());
-                this.methodHandle = MethodWrapper.LOOKUP.findVirtual(nodeClass, m.getName(), methodType).asType(GETTER_TYPE);
-            } else {
-                this.methodHandle = LOOKUP.unreflect(m).asType(GETTER_TYPE);
-            }
+            // Note: We only support public methods on public types. If the method being called is implemented
+            // in a package-private class, this won't work.
+            // See git history here and https://github.com/pmd/pmd/issues/4885
+            this.methodHandle = LOOKUP.unreflect(m).asType(GETTER_TYPE);
         }
 
 

--- a/pmd-core/src/test/java/net/sourceforge/pmd/lang/rule/xpath/impl/dummyast/AbstractNode.java
+++ b/pmd-core/src/test/java/net/sourceforge/pmd/lang/rule/xpath/impl/dummyast/AbstractNode.java
@@ -8,16 +8,16 @@ import net.sourceforge.pmd.lang.ast.DummyNode;
 import net.sourceforge.pmd.lang.document.Chars;
 
 // This class is package private
-// and provides the implementation for getValue(). This
-// class is the DeclaringClass for that method.
-class AbstractNode extends DummyNode implements ValueNode {
+// and provides the implementation for getValue().
+// This method is not accessible from outside this package,
+// it is made available in the subclass ConcreteNode.
+class AbstractNode extends DummyNode {
 
     AbstractNode() {
 
     }
 
-    @Override
-    public final Chars getValue() {
+    Chars getValue() {
         return Chars.wrap("actual_value");
     }
 }

--- a/pmd-core/src/test/java/net/sourceforge/pmd/lang/rule/xpath/impl/dummyast/ConcreteNode.java
+++ b/pmd-core/src/test/java/net/sourceforge/pmd/lang/rule/xpath/impl/dummyast/ConcreteNode.java
@@ -4,5 +4,11 @@
 
 package net.sourceforge.pmd.lang.rule.xpath.impl.dummyast;
 
+import net.sourceforge.pmd.lang.document.Chars;
+
 public final class ConcreteNode extends AbstractNode implements ValueNode {
+    @Override
+    public Chars getValue() {
+        return super.getValue();
+    }
 }

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTBooleanLiteral.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTBooleanLiteral.java
@@ -6,10 +6,12 @@ package net.sourceforge.pmd.lang.java.ast;
 
 import org.checkerframework.checker.nullness.qual.NonNull;
 
+import net.sourceforge.pmd.lang.document.Chars;
+
 /**
  * The boolean literal, either "true" or "false".
  */
-public final class ASTBooleanLiteral extends AbstractLiteral {
+public final class ASTBooleanLiteral extends AbstractLiteral implements ASTLiteral {
 
     private boolean isTrue;
 
@@ -30,6 +32,11 @@ public final class ASTBooleanLiteral extends AbstractLiteral {
     @Override
     public @NonNull Boolean getConstValue() {
         return isTrue;
+    }
+
+    @Override
+    public Chars getLiteralText() {
+        return super.getLiteralText();
     }
 
     @Override

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTCharLiteral.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTCharLiteral.java
@@ -15,7 +15,7 @@ import net.sourceforge.pmd.lang.document.Chars;
  * retrieve the actual runtime value. Use {@link #getLiteralText()} to
  * retrieve the text.
  */
-public final class ASTCharLiteral extends AbstractLiteral {
+public final class ASTCharLiteral extends AbstractLiteral implements ASTLiteral {
 
 
     ASTCharLiteral(int id) {
@@ -39,4 +39,8 @@ public final class ASTCharLiteral extends AbstractLiteral {
         return StringEscapeUtils.UNESCAPE_JAVA.translate(woDelims).charAt(0);
     }
 
+    @Override
+    public Chars getLiteralText() {
+        return super.getLiteralText();
+    }
 }

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTNullLiteral.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTNullLiteral.java
@@ -6,6 +6,8 @@ package net.sourceforge.pmd.lang.java.ast;
 
 import org.checkerframework.checker.nullness.qual.Nullable;
 
+import net.sourceforge.pmd.lang.document.Chars;
+
 /**
  * The null literal.
  *
@@ -15,7 +17,7 @@ import org.checkerframework.checker.nullness.qual.Nullable;
  *
  * </pre>
  */
-public final class ASTNullLiteral extends AbstractLiteral {
+public final class ASTNullLiteral extends AbstractLiteral implements ASTLiteral {
     ASTNullLiteral(int id) {
         super(id);
     }
@@ -34,5 +36,10 @@ public final class ASTNullLiteral extends AbstractLiteral {
     @Override
     public @Nullable Object getConstValue() {
         return null;
+    }
+
+    @Override
+    public Chars getLiteralText() {
+        return super.getLiteralText();
     }
 }

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTNumericLiteral.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTNumericLiteral.java
@@ -14,7 +14,7 @@ import net.sourceforge.pmd.lang.java.types.JPrimitiveType;
 /**
  * A numeric literal of any type (double, int, long, float, etc).
  */
-public final class ASTNumericLiteral extends AbstractLiteral {
+public final class ASTNumericLiteral extends AbstractLiteral implements ASTLiteral {
 
     /**
      * True if this is an integral literal, ie int OR long,
@@ -34,6 +34,11 @@ public final class ASTNumericLiteral extends AbstractLiteral {
     @Override
     protected <P, R> R acceptVisitor(JavaVisitor<? super P, ? extends R> visitor, P data) {
         return visitor.visit(this, data);
+    }
+
+    @Override
+    public Chars getLiteralText() {
+        return super.getLiteralText();
     }
 
     @Override

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTStringLiteral.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTStringLiteral.java
@@ -18,7 +18,7 @@ import net.sourceforge.pmd.util.StringUtil;
  * in the source ({@link #getLiteralText()}). {@link #getConstValue()} allows to recover
  * the actual runtime value, by processing escapes.
  */
-public final class ASTStringLiteral extends AbstractLiteral {
+public final class ASTStringLiteral extends AbstractLiteral implements ASTLiteral {
 
     private static final String TEXTBLOCK_DELIMITER = "\"\"\"";
 
@@ -34,6 +34,11 @@ public final class ASTStringLiteral extends AbstractLiteral {
     @Override
     public String getImage() {
         return getText().toString();
+    }
+
+    @Override
+    public Chars getLiteralText() {
+        return super.getLiteralText();
     }
 
     void setTextBlock() {

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractLiteral.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractLiteral.java
@@ -10,8 +10,11 @@ import net.sourceforge.pmd.lang.rule.xpath.NoAttribute;
 
 /**
  * @author Clément Fournier
+ * @see ASTLiteral#getLiteralText()
+ * @see #getLiteralText()
  */
-abstract class AbstractLiteral extends AbstractJavaExpr implements ASTLiteral {
+// Note: This class must not implement ASTLiteral, see comment on #getLiteralText()
+abstract class AbstractLiteral extends AbstractJavaExpr {
 
     private JavaccToken literalToken;
 
@@ -41,12 +44,15 @@ abstract class AbstractLiteral extends AbstractJavaExpr implements ASTLiteral {
         return firstToken.getImageCs();
     }
 
-    @Override
-    public final Chars getLiteralText() {
+    // This method represents ASTLiteral#getLiteralText().
+    // However, since this class is package private, this method is not reliably accessible
+    // via reflection/method handles (see https://github.com/pmd/pmd/issues/4885).
+    // Subclasses of this class need to implement ASTLiteral and override this method
+    // as public.
+    Chars getLiteralText() {
         assert literalToken.getImageCs() != null;
         return literalToken.getImageCs();
     }
-
 
     @Override
     public boolean isCompileTimeConstant() {

